Chown contents on volume mount if user specifies :u

rhatdan opened the pull request:
This patch follows the same pattern as the :Z and :z patch.

Basically it will recursively chown the content on container
creation iff user specifies :u on volume mount and --user at the
same time.  Once we get usernamespace this could be extended to
do the same thing to match the primary UID of the container.

Docker-DCO-1.1-Signed-off-by: Dan Walsh dwalsh@redhat.com (github: rhatdan)

pmorie commented:
Looks like this solves #7198, does it not, @rhatdan?

rhatdan commented:
@pmorie That is the idea.

pmorie commented:
@rhatdan Will the chown result be visible from the host?

rhatdan commented:
yes

thaJeztah commented:
Hm, sounds dangerous, thinking of people reporting "I ran `docker run -v /:/foo:u --user foobar ubuntu`, and now my host acts funny"

pmorie commented:
@thaJeztah I definitely think it deserves a lot of doc and a big fat warning, just like `--privileged`.

rhatdan commented:
@thaJeztah I agree, we have checks in the SELinux one to at least block the most common examples.

/ /etc /usr

cpuguy83 commented:
I don't see the purpose of this.

1. We should not ever be changing the contents of bind mounts. These can be changed by the operator as needed.
2. When using a normal volume, the volume is already chowned to the owner of those files from the container fs.

rhatdan commented:
I believe the use case is for containers with UIDs built in.  IE If I build a container to run as USER 512 group 1234, and all the internal content has to be owned by UID 512, GID 1234, it allows for easier use of volume mounts into the container.  I agree you could somehow wrap the container with something that would inspect the image UID/GID and then do the chown externally before doing the volume mounts.

rhatdan commented:
But a lot of the discussion of this has happened on https://github.com/docker/docker/issues/7198

pmorie commented:
> I believe the use case is for containers with UIDs built in. IE If I build a container to run as USER 512 group 1234, and all the internal content has to be owned by UID 512, GID 1234, it allows for easier use of volume mounts into the container. I agree you could somehow wrap the container with something that would inspect the image UID/GID and then do the chown externally before doing the volume mounts

@rhatdan @cpuguy83 

That is exactly right.  The use-case from Kubernetes / OpenShift is that we need to be able to provide volumes to pods running as non-zero uid/gid and have the uid/gid on the volumes match the effective uid/gid of the container.  Right now, in order to consume volumes properly from a container running this way, you have to do a hack like setting up `nss_wrapper`, which is not a great experience for users.  Fundamentally this is about allowing non-root processes to easily consume volumes.

thaJeztah commented:
@pmorie in that use case, are the volumes bind-mounted, or "regular" volumes?

pmorie commented:
@thaJeztah They are bind-mounted

phemmer commented:
Just throwing in my $0.02, I don't like this feature as it breaks the concept that containers don't affect the host system.

Also, what happens if you have 2 containers volume mount the same path, but they have different user IDs?

I think something such as this is best done as either:
1) Left up to the user to manage.
2) Create a FUSE driver to act as a filesystem proxy.

pmorie commented:
@phemmer 

In the use-case we have in mind, if different containers mount the same volume they will necessarily be running as the same uid/gid

calavera commented:
If I understand the thread in the original issue and here, this can be solved by making the privileges in the host slightly more open. I don't think docker should change any privileges in the host, it should be up to the operator.

pmorie commented:
@calavera In some use-cases, relaxing the permissions won't work.  For example, many packages don't like having 0777 privileges -- we need to be able to support 0770 with the right uid/gid

Let me also throw out there that my personal opinion is that this is much easier to do the right way with the user ns, and I think long-term that will be the best way to go.

rhatdan commented:
I actually don't see how this helps with the user namespace, I believe the User Namespace is actually going to cause more of these problems.  

This all stems from the fact that a file object can only have a single UID associated with it. (Unless we entered the realm of ACLs).

If I have a container that runs as UID 1234 and has all of the objects in the image owned by UID 1234, how does UserNamespace help with the fact that I mount a volume into it owned by UID 0?

pmorie commented:
@rhatdan It's entirely possible that I'm wrong :)

Specifically what I had in mind was to use a random uid/gid for the volume
on a container and remap to effective uid/gid in container.  It doesn't buy
you anything from what I can tell if you're mapping _from_ 0:0 on host, but
if you make arbitrary non-root uid/gid on host, then you can remap however
you please.

Let me know if I'm wrong!

On Wed, Jul 15, 2015 at 2:02 PM, Daniel J Walsh <notifications@github.com>
wrote:

> I actually don't see how this helps with the user namespace, I believe the
> User Namespace is actually going to cause more of these problems.
>
> This all stems from the fact that a file object can only have a single UID
> associated with it. (Unless we entered the realm of ACLs).
>
> If I have a container that runs as UID 1234 and has all of the objects in
> the image owned by UID 1234, how does UserNamespace help with the fact that
> I mount a volume into it owned by UID 0?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14632#issuecomment-121694219>.
>

thaJeztah commented:
Really thinking out loud here (as I haven't looked good enough into what's possible there), but now that we have volume-driver plugins, could this be implemented as a plugin?

rhatdan commented:
@thaJeztah No idea, I also have not looked at these plugins.

@pmorie The fundamental problem with user namespace is lack of file system support.

Lets say you have all content in a container owned by 1234.  Now I user 5678 to use the container.  I also want to volume mount my home directory into the container which has content owned by 5678.  If I map UID 5678 to 1234 and run the container, then the processes inside the container will be running as 1234, and will have full access to my homedir, since the files in my homedir as owned by real UID 5678.  BUT the content inside the container is owned by 1234, since uid 1234 is not mapped into the container then this content will be treated as UID -1 and will not be accessible to my processes.  (AS I understand it). w

eparis commented:
This is exactly as dangerous as the selinux :z feature no doubt about it. Exactly the same way that :z can add or remove permissions so can :u.  Blocking :u in the same places we block :z makes perfect sense.

Consider `docker run --user=bob -v /tmp/container1:/tmp`  the host operator actually has no/extremely little idea what the required uid is.  the "bob" uid is determined by /etc/passwd inside the container, right? This is the exact same reason that :z is required. The operator doesn't know what value docker is going to use, and in fact this value could change! The operator is not able to do the chown -R himself with ANY degree of usefulness.

However `docker run --user=bob -b /tmp/container1:/tmp:u,z` does in fact work, is predictable and reliable, and provides the expected user experience.

pmorie commented:
@rhatdan Okay, you have convinced me that I have a lot of reading to do re: user namespace.  It looks from some initial googling like there will probably be per-fs issues around user ns to work through or that will differentiate how successful you can use userns with different file systems.  (context: I wound up looking through [this link](http://thread.gmane.org/gmane.linux.kernel.containers/25169))

rhatdan commented:
The problem with User Namespace is almost no one has used it yet, so everyone is making assumptions on what it will do, and a lot of those assumptions are totally wrong.  Hopefully we can get it turned on in Docker soon, so people can start experimenting and figuring out the best use cases.   But it is certainly not the Nirvana that everyone imagines it to be.

cpuguy83 commented:
It's going to be awful until there is some FS level support...

pmorie commented:
@rhatdan 

> The problem with User Namespace is almost no one has used it yet, so everyone is making assumptions on what it will do, and a lot of those assumptions are totally wrong. Hopefully we can get it turned on in Docker soon, so people can start experimenting and figuring out the best use cases.

Yeah, I was thinking about this over the weekend and wondered what the easiest way to start experimenting with it would be.  Is there a branch where the work is happening?  Is there a timeline around it appearing in a docker release?

rhatdan commented:
Any chance of moving this forward.  It really is necessary for the k8s folks.  I updated the pull request to stop chown of /, /etc, /usr to make the Relabel code in Z,z.

LK4D4 commented on daemon/volumes.go:
```diff
@@ -226,12 +226,15 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
+		if err := label.Relabel(bind.Source, container.MountLabel, bind.Mode); err != nil {
+			return err
+		}
+		if err := volume.FixUidGid(container.Config.User, bind.Source, bind.Mode); err != nil {
```
So, if I understand right this is going on in host mount namespace. And if there is no such user on host(which is likely, it can also have another uid/gid) chown will be done to current user of docker daemon(most likely root). TLDR: I don't understand what this patch do.

rhatdan replied:
It will be dealing with a UID not a username, so the user not existing on the host will not matter.  The usual use case for this is for containers that are not running as root.

LK4D4 replied:
Why it will be dealing with uid? `container.Config.User` is a string and it can specify username, gid, groupname and stuff.

LK4D4 commented:
I posted question about implementation. Seems weird for me.

LK4D4 commented on volume/volume.go:
```diff
@@ -25,33 +35,93 @@ type Volume interface {
+		Gid:  syscall.Getgid(),
+		Home: "/",
+	}
+	passwdPath, err := user.GetPasswdPath()
```
Just trying to understand patch. You probably want this files from inside rootfs, not from host?

rhatdan replied:
This is the exact code copied from the handling of -u when processes are execed in libcontainer.  It reads the /etc/passwd from inside the container to determine the UID/GID of the container.  I believe it will get the correct UID/GID based on the container contents.

LK4D4 replied:
btw I forgot to comment, but no it won't get correct uid/gid. libcontainer code executed after pivot_root

tiborvass commented:
*Collective review* @vbatts @cpuguy83 @duglin @abronan 

We should probably wait until user namespaces are finalized before this is in master, to have a better grasp on how we'd be able to extend the chown-ing behavior to use user/mount namespaces.

However, we're fine with this opt-in behavior if it's part of the experimental channel.
@rhatdan does that sound okay to you?

rhatdan commented:
@tiborvass It is fine with me. But I am just the conduit of needs from the k8s community. @eparis What do  you think?

rhatdan committed:
Chown contents on volume mount if user specified :u

This patch follows the same pattern as the :Z and :z patch.

Basically it will recursively chown the content on container
creation iff user specifies :u on volume mount and --user at the
same time.  Once we get usernamespace this could be extended to
do the same thing to match the primary UID of the container.

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

Dan Walsh <dwalsh@redhat.com>

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

tiborvass commented:
@eparis would that be acceptable for this PR to add this feature to the experimental channel instead of stable?

eparis commented:
I'd be ok if it was experimental.  Usernamespace is going to make everything hard   :)

tiborvass commented:
@eparis thanks :)

@rhatdan feel free to update the code to make it part of experimental.

rhatdan commented:
Sure, although I really do not know what I need to do?  Do you have any docs on making an option experimental?

thaJeztah commented:
@rhatdan I think it's all handled via build flags, e.g. see https://github.com/docker/docker/blob/master/daemon/config_experimental.go

rhatdan commented:
Except I am not adding a new option but extending the volume option.

tiborvass commented:
Somebody please help @rhatdan with porting the code to experimental.

Just FYI, user namespaces will make this harder :P

thaJeztah commented:
I asked @runcom to assist in moving the changes to experimental (thanks Antonio!)

runcom committed:
move chown contents on volume with :u to experimental

Signed-off-by: Antonio Murdaca <amurdaca@redhat.com>

rhatdan committed:
Merge pull request #122 from runcom/pr/14632

move chown contents on volume with :u to experimental

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "user" git@github.com:rhatdan/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

runcom commented:
closing since I carried it here https://github.com/docker/docker/pull/16767

rhatdan commented:
SGTM
